### PR TITLE
Feat/RSS-ECOMM-2_01: Implement client-side validation for login form

### LINKED ISSUE #58: RSS-ECOMM-2_01: Implement client-side validation for login form (20 points)
Implement client-side validation for the login form, including email 📧 and password 🔑 fields. 
Ensure that input validation checks are performed in real-time when the user enters their information.

**Email Validation Rules** 📏
- [x] Email address must be properly formatted (e.g., [user@example.com](mailto:user@example.com)).
- [x] Email address must not contain leading or trailing whitespace.
- [x] Email address must contain a domain name (e.g., example.com).
- [x] Email address must contain an '@' symbol separating local part and domain name.
**Password Validation Rul** 📏
- [x] Password must be at least 8 characters long.
- [x] Password must contain at least one uppercase letter (A-Z).
- [x] Password must contain at least one lowercase letter (a-z).
- [x] Password must contain at least one digit (0-9).
- [x] (Optional) Password must contain at least one special character (e.g., !@#$%^&*).
- [x] Password must not contain leading or trailing whitespace.
**Show/Hide Password Functionality** 👁️‍🗨️
- [x] Provide a button or a checkbox to allow users to toggle the visibility of their password.
- [x] When this option is activated, users should be able to see their password in plaintext.
- [x] When this option is deactivated, the password should be replaced with asterisks or bullet points.
**Acceptance Criteria** ✅
- [x] Email field validation checks for a properly formatted email address according to the specified rules.
- [x] Password field validation checks for meeting the length and complexity requirements.
- [x] Show/hide password functionality allows users to toggle password visibility.
- [ ] Validation checks are performed as the user types.

### LINKED ISSUE #59: RSS-ECOMM-2_02: Display clear error messages for validation issues (20 points) 🚨
Display clear error messages on the login form when validation issues are detected, such as an improperly formatted email address 📧 or a weak password 🔑. The error messages should follow specific display rules for a consistent and user-friendly experience.

**Display Rules for Validation Error Messages** 📚
- [x] Error messages should appear next to the corresponding input field (e.g., either to the right, left, top, or bottom of the input field).
- [x] Error messages should be styled consistently (e.g., with a specific font, color, and size) across all input fields.
- [x] Error messages should not overlap other UI elements or obstruct any important information.
- [x] Error messages should disappear once the user corrects the invalid input and meets the validation criteria.
- [x] Error messages should be easily readable and visually distinct from other text on the page (e.g., by using a different color or font weight).
**Acceptance Criteria** ✅
- [x] Clear error messages are shown next to the corresponding input field when validation fails.
- [x] Error messages are descriptive, explaining the specific validation issue.
- [x] Error messages follow the specified display rules for a consistent and user-friendly experience.
- [x] Error messages disappear when the user corrects the invalid input.

*So team, let's take a leaf from Pink Floyd's book and remember - "All in all, it's just another brick in the wall." And every error message we solve is one less brick in our user's wall! Let's bring the wall down together!* 🎸🧱

### PR DESCRIPTION
**Pull request by Tetiana-KET**
RSS-ECOMM-2_01: Implement client-side validation for login form
resolve #58, resolve #59

#### 🤔 This is a ...

- [x] 🌟 New task

#### Description

Implement client-side validation for the login form, including email 📧 and password 🔑 fields. 
Ensure that input validation checks are performed in real-time when the user enters their information.

**Email Validation Rules** 📏
- [x] Describe tests for Email Validation 
- [x] Email address must be properly formatted (e.g., [user@example.com](mailto:user@example.com)).
- [x] Email address must not contain leading or trailing whitespace.
- [x] Email address must contain a domain name (e.g., example.com).
- [x] Email address must contain an '@' symbol separating local part and domain name.
**Password Validation Rul** 📏
- [x] Describe tests for Password Validation 
- [x] Password must be at least 8 characters long.
- [x] Password must contain at least one uppercase letter (A-Z).
- [x] Password must contain at least one lowercase letter (a-z).
- [x] Password must contain at least one digit (0-9).
- [x] Password must contain at least one special character (e.g., !@#$%^&*).
- [x] Password must not contain leading or trailing whitespace.
**Show/Hide Password Functionality** 👁️‍🗨️
- [x] Provide a button or a checkbox to allow users to toggle the visibility of their password.
- [x] When this option is activated, users should be able to see their password in plaintext.
- [x] When this option is deactivated, the password should be replaced with asterisks or bullet points.
**Acceptance Criteria** ✅
- [x] Email field validation checks for a properly formatted email address according to the specified rules.
- [x] Password field validation checks for meeting the length and complexity requirements.
- [x] Show/hide password functionality allows users to toggle password visibility.
- [x] Validation checks are performed as the user types.

#### Checklist

- [x] ✅ I have performed a self-review of my own code.
- [x] 📝 I have commented my code, particularly in hard-to-understand areas.
- [x] 🔧 I have made corresponding changes to the documentation (if applicable).
- [x] 🚫 My changes generate no new warnings or errors.
- [x] 👌 Have at least 1 approve.
- [x] 🌳 Branch from `develop`/`release/...` and target to `develop`/`release/...`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,12 +29,13 @@
     "@mui/icons-material": "^5.15.16",
     "@mui/material": "^5.15.16",
     "@mui/x-date-pickers": "^7.3.2",
-    "react": "^18.2.0",
-    "react-dom": "^18.2.0",
+    "react": "^17.0.0 || ^18.0.0",
```
Comment (JsPowWow) [inline package.json]: Revert this two lines as it was before

File: src/pages/Login/schema.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+const passwordValidationRegEx = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]{8,20}$/;
+const invalidPasswordMessage =
+  'Passwords must consist only of Latin letters (A-Z, a-z), be 8 - 20 characters and contain at least: one uppercase letter (A-Z), one lowercase letter (a-z), one digit (0-9) and one special character (e.g., !@#$%^&*)';
+
+export const loginFormSchema = z.object({
+  email: z.string({ required_error: 'Email is required' }).email('Invalid email'),
+  password: z
+    .string()
+    .min(MIN_LENGTH, 'Password must be at least 8 characters long')
+    .max(MAX_LENGTH, 'Password should not be more, then 20 characters')
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]: Not 8, but `...${MIN_LENGTN}...`
Not 20, but `...${MAX_LENGTN}...`
Use Your constants in messages

File: src/pages/Login/schema.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { z } from 'zod';
+
+const MIN_LENGTH = 8;
+const MAX_LENGTH = 20;
+const passwordValidationRegEx = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]{8,20}$/;
+const invalidPasswordMessage =
+  'Passwords must consist only of Latin letters (A-Z, a-z), be 8 - 20 characters and contain at least: one uppercase letter (A-Z), one lowercase letter (a-z), one digit (0-9) and one special character (e.g., !@#$%^&*)';
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]: Not 8, but `...${MIN_LENGTN}...`
Not 20, but `...${MAX_LENGTN}...`
Use Your constants in messages

File: src/pages/Login/schema.ts (hunk 2 above)
Comment (Tetiana-KET) [reply src/pages/Login/schema.ts]: done, thank you

File: src/pages/Login/schema.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { z } from 'zod';
+
+export const MIN_LENGTH = 8;
+export const MAX_LENGTH = 20;
+const passwordValidationRegEx = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]{8,20}$/;
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]:
```suggestion
const TEST_PASSWORD_REG_EX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]{8,20}$/;
```

File: src/pages/Login/schema.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  password: z
+    .string()
+    .min(MIN_LENGTH, `Password must be at least ${MIN_LENGTH} characters long`)
+    .max(MAX_LENGTH, `Password should not be more, then ${MAX_LENGTH} characters`)
+    .regex(/^[^\s]+$/, 'Passwords must not contain whitespaces')
+    .regex(/(?=.*[a-z])(?=.*[A-Z])/, 'Passwords must contain uppercase and lowercase Latin letters (A-Z, a-z)')
+    .regex(/(?=.*\d)/, 'Passwords must contain at least one digit (0-9)')
+    .regex(/(?=.*[!@#$%^&])/, 'Passwords must contain at least one special character (!@#$%^&)')
+    .regex(passwordValidationRegEx, 'Invalid password'),
+});
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]:
```suggestion
    .regex(TEST_PASSWORD_REG_EX, 'Invalid password'),
```

File: src/pages/Login/Login.test.tsx [hunk 6]
```diff
@@ -1,8 +1,173 @@
...
+
+  // Passwords must consist only of Latin letters (A-Z, a-z)
+  test('should return false if password contains not only Latin letters', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'абвгд@R%id&12!#$*' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Invalid password');
+  });
+
+  // test valid password
+  test('should return true if password is valid', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password should be valid', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 7]
```diff
@@ -1,8 +1,173 @@
...
+  });
+
+  test('should return false if password contains whitespace', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'V@al% id&12!#$^* ' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Passwords must not contain whitespaces');
+  });
+
+  // Passwords must consist only of Latin letters (A-Z, a-z)
+  test('should return false if password contains not only Latin letters', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password must consist only of Latin letters (A-Z, a-z)', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 8]
```diff
@@ -1,8 +1,173 @@
...
+    expect(result.error?.issues[0].message).toBe('Passwords must not contain whitespaces');
+  });
+
+  test('should return false if password contains trailing whitespace', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'V@al%id&12!#$^* ' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Passwords must not contain whitespaces');
+  });
+
+  test('should return false if password contains whitespace', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password should not contains whitespace', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 9]
```diff
@@ -1,8 +1,173 @@
...
+  });
+
+  // Password must not contain leading or trailing or middle whitespace.
+  test('should return false if password contains leading whitespace', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: ' V@al%id&12!#$^*' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Passwords must not contain whitespaces');
+  });
+
+  test('should return false if password contains trailing whitespace', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password should not contains trailing whitespace', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 10]
```diff
@@ -1,8 +1,173 @@
...
+
+  // Password must contain at least one special character (e.g., !@#$%^&*).
+  test('should return false if password does not contain special character', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'inValidPass12' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Passwords must contain at least one special character (!@#$%^&)');
+  });
+
+  // Password must not contain leading or trailing or middle whitespace.
+  test('should return false if password contains leading whitespace', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password must not contain leading whitespace', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 11]
```diff
@@ -1,8 +1,173 @@
...
+
+  // Password must contain at least one digit (0-9)
+  test('should return false if password does not contain digits', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'InvalidPass' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Passwords must contain at least one digit (0-9)');
+  });
+
+  // Password must contain at least one special character (e.g., !@#$%^&*).
+  test('should return false if password does not contain special character', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password must contain at least one special character', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 12]
```diff
@@ -1,8 +1,173 @@
...
+  test('should return false if password does not contain letters', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: '12345678' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe(
+      'Passwords must contain uppercase and lowercase Latin letters (A-Z, a-z)',
+    );
+  });
+
+  // Password must contain at least one digit (0-9)
+  test('should return false if password does not contain digits', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password must contain at least one digit', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 13]
```diff
@@ -1,8 +1,173 @@
...
+  // Password must contain at least one lowercase letter (a-z)
+  test('should return false if password does not contain at least one lowercase letter', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'INVALID5414' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe(
+      'Passwords must contain uppercase and lowercase Latin letters (A-Z, a-z)',
+    );
+  });
+
+  test('should return false if password does not contain letters', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password must contain at least one uppercase & lowercase Latin letters', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 14]
```diff
@@ -1,8 +1,173 @@
...
+  test('should return false if password does not contain at least one uppercase letter', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'invalid5414' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe(
+      'Passwords must contain uppercase and lowercase Latin letters (A-Z, a-z)',
+    );
+  });
+
+  // Password must contain at least one lowercase letter (a-z)
+  test('should return false if password does not contain at least one lowercase letter', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('Password must contain at least one lowercase Latin letter', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 15]
```diff
@@ -1,8 +1,173 @@
...
+
+  // Password should not be more, then 20 characters
+  test(`should return false if password length is more then ${MAX_LENGTH}`, () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: 'inValidPass5416823548' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe(`Password should not be more, then ${MAX_LENGTH} characters`);
+  });
+
+  // Password must contain at least one uppercase letter (A-Z)
+  test('should return false if password does not contain at least one uppercase letter', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password must contain at least one uppercase letter', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 16]
```diff
@@ -1,8 +1,173 @@
...
+
+  // Password must be at least 8 characters long
+  test(`should return false if password length is less them ${MIN_LENGTH}`, () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: '54168' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe(`Password must be at least ${MIN_LENGTH} characters long`);
+  });
+
+  // Password should not be more, then 20 characters
+  test(`should return false if password length is more then ${MAX_LENGTH}`, () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test(`password length should not be more, then 20 characters`, () => {
```

File: src/pages/Login/Login.test.tsx [hunk 17]
```diff
@@ -1,8 +1,173 @@
...
+describe('password validation', () => {
+  // Password must not be an empty line
+  test('should return false if password is empty', () => {
+    const result = loginFormSchema.safeParse({ email: 'correct_email@gmail.com', password: '' });
+    expect(result.error?.issues[0].message).toBe(`Password must be at least ${MIN_LENGTH} characters long`);
+    expect(result.success).toBe(false);
+  });
+
+  // Password must be at least 8 characters long
+  test(`should return false if password length is less them ${MIN_LENGTH}`, () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test(`password length must be at least 8 characters long`, () => {
```

**Review by JsPowWow — CHANGES_REQUESTED** [review]

File: src/pages/Login/Login.test.tsx [hunk 18]
```diff
@@ -1,8 +1,173 @@
...
+
+  test('should return true when email is valid', () => {
+    const result = loginFormSchema.safeParse({ email: 'validemail@gmail.com', password: 'Valid@al%id&12!#$^*' });
+    expect(result.success).toBe(true);
+  });
+});
+
+describe('password validation', () => {
+  // Password must not be an empty line
+  test('should return false if password is empty', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('password must not be empty', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 19]
```diff
@@ -1,8 +1,173 @@
...
+    const result = loginFormSchema.safeParse({ email: 'valid-email@gmail.com', password: 'Valid@al%id&12!#$^*' });
+    expect(result.success).toBe(true);
+  });
+
+  test('should return true when email contains "."', () => {
+    const result = loginFormSchema.safeParse({ email: 'valid.email@gmail.com', password: 'Valid@al%id&12!#$^*' });
+    expect(result.success).toBe(true);
+  });
+
+  test('should return true when email is valid', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email should be valid', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 20]
```diff
@@ -1,8 +1,173 @@
...
+    const result = loginFormSchema.safeParse({ email: 'valid_email@gmail.com', password: 'Valid@al%id&12!#$^*' });
+    expect(result.success).toBe(true);
+  });
+
+  test('should return true when email contains "-"', () => {
+    const result = loginFormSchema.safeParse({ email: 'valid-email@gmail.com', password: 'Valid@al%id&12!#$^*' });
+    expect(result.success).toBe(true);
+  });
+
+  test('should return true when email contains "."', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email with "." should be valid', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 21]
```diff
@@ -1,8 +1,173 @@
...
+    expect(result.error?.issues[0].message).toBe('Invalid email');
+  });
+
+  // test valid email provided, Email address may contain "-", "_", ".".
+  test('should return true when email contains "_"', () => {
+    const result = loginFormSchema.safeParse({ email: 'valid_email@gmail.com', password: 'Valid@al%id&12!#$^*' });
+    expect(result.success).toBe(true);
+  });
+
+  test('should return true when email contains "-"', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email with "-" should be valid', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 22]
```diff
@@ -1,8 +1,173 @@
...
+  });
+
+  test('should return false if email contains trailing whitespace', () => {
+    const result = loginFormSchema.safeParse({ email: 'codjhwdj@gmail.com ', password: '5123456784168' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Invalid email');
+  });
+
+  // test valid email provided, Email address may contain "-", "_", ".".
+  test('should return true when email contains "_"', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email with "_" should be valid', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 23]
```diff
@@ -1,8 +1,173 @@
...
+  });
+
+  // Email address must not contain leading or trailing whitespace.
+  test('should return false if email contains leading whitespace', () => {
+    const result = loginFormSchema.safeParse({ email: ' codjhwdj@gmail.com', password: '5123456784168' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Invalid email');
+  });
+
+  test('should return false if email contains trailing whitespace', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email should not contains trailing whitespace', () => {
```

**Review by JsPowWow — CHANGES_REQUESTED** [review]

File: src/pages/Login/Login.test.tsx [hunk 24]
```diff
@@ -1,8 +1,173 @@
...
+
+  //  Email address must contain an '@' symbol separating local part and domain name.
+  test('should return false if email does not contain "@"', () => {
+    const result = loginFormSchema.safeParse({ email: 'user.example.com', password: '5416854g64s5g4' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Invalid email');
+  });
+
+  // Email address must not contain leading or trailing whitespace.
+  test('should return false if email contains leading whitespace', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email should not contain leading whitespace', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 25]
```diff
@@ -1,8 +1,173 @@
...
+  });
+
+  test('should return false if email does not contain a domain name', () => {
+    const result = loginFormSchema.safeParse({ email: 'user@', password: '5416854g64s5g4' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Invalid email');
+  });
+
+  //  Email address must contain an '@' symbol separating local part and domain name.
+  test('should return false if email does not contain "@"', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email should contain "@"', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 26]
```diff
@@ -1,8 +1,173 @@
...
+  });
+
+  // Email address must be properly formatted and contain a domain name (e.g., user@example.com)
+  test('should return false if email is not properly formatted', () => {
+    const result = loginFormSchema.safeParse({ email: 'useremail', password: '5416854g64s5g4' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Invalid email');
+  });
+
+  test('should return false if email does not contain a domain name', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email should contain a domain name', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 27]
```diff
@@ -1,8 +1,173 @@
...
+describe('email validation', () => {
+  // Email address must not be an empty string
+  test('should return false if email is an empty string', () => {
+    const result = loginFormSchema.safeParse({ email: '', password: '5416854g64s5g4' });
+    expect(result.success).toBe(false);
+    expect(result.error?.issues[0].message).toBe('Invalid email');
+  });
+
+  // Email address must be properly formatted and contain a domain name (e.g., user@example.com)
+  test('should return false if email is not properly formatted', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email address must be properly formatted and contain a domain name', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 28]
```diff
@@ -1,8 +1,173 @@
...
+
+test('should return false if receives an empty object', () => {
+  const result = loginFormSchema.safeParse({});
+  expect(result.success).toBe(false);
+  expect(result.error?.issues[0].message).toBe('Email is required');
+});
+
+describe('email validation', () => {
+  // Email address must not be an empty string
+  test('should return false if email is an empty string', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
  test('email address must not be empty', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 29]
```diff
@@ -1,8 +1,173 @@
...
   expect(true).toBeTruthy();
 });
+
+test('should return false if receives an empty object', () => {
+  const result = loginFormSchema.safeParse({});
+  expect(result.success).toBe(false);
+  expect(result.error?.issues[0].message).toBe('Email is required');
+});
+
+describe('email validation', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
describe('email address validation tests', () => {
```

File: src/pages/Login/Login.test.tsx [hunk 30]
```diff
@@ -1,8 +1,173 @@
...
 import { render } from '@testing-library/react';
+import { MAX_LENGTH, MIN_LENGTH, loginFormSchema } from './schema';
 import LoginPage from './Login.page';
 
 test('Render the login page', () => {
   render(<LoginPage />);
   expect(true).toBeTruthy();
 });
+
+test('should return false if receives an empty object', () => {
```
Comment (JsPowWow) [inline src/pages/Login/Login.test.tsx]:
```suggestion
test('email address should be provided', () => {
```

File: src/pages/Login/schema.ts [hunk 31]
```diff
@@ -0,0 +1,18 @@
...
+export const MAX_LENGTH = 20;
+const passwordValidationRegEx = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]{8,20}$/;
+
+export const loginFormSchema = z.object({
+  email: z.string({ required_error: 'Email is required' }).email('Invalid email'),
+  password: z
+    .string()
+    .min(MIN_LENGTH, `Password must be at least ${MIN_LENGTH} characters long`)
+    .max(MAX_LENGTH, `Password should not be more, then ${MAX_LENGTH} characters`)
+    .regex(/^[^\s]+$/, 'Passwords must not contain whitespaces')
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]:
```suggestion
    .regex(/^[^\s]+$/, 'Password must not contain whitespaces')
```

File: src/pages/Login/schema.ts [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+    .regex(/(?=.*[a-z])(?=.*[A-Z])/, 'Passwords must contain uppercase and lowercase Latin letters (A-Z, a-z)')
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]:
```suggestion
    .regex(/(?=.*[a-z])(?=.*[A-Z])/, 'Password must contain uppercase and lowercase Latin letters (A-Z, a-z)')
```

File: src/pages/Login/schema.ts [hunk 33]
```diff
@@ -0,0 +1,18 @@
...
+    .regex(/(?=.*\d)/, 'Passwords must contain at least one digit (0-9)')
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]:
```suggestion
    .regex(/(?=.*\d)/, 'Password must contain at least one digit (0-9)')
```

File: src/pages/Login/schema.ts [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+    .regex(/(?=.*[!@#$%^&])/, 'Passwords must contain at least one special character (!@#$%^&)')
```
Comment (JsPowWow) [inline src/pages/Login/schema.ts]:
```suggestion
    .regex(/(?=.*[!@#$%^&])/, 'Password must contain at least one special character (!@#$%^&)')
```

**Review by JsPowWow — CHANGES_REQUESTED** [review]

File: src/pages/Login/schema.ts (hunk 5 above)
Comment (Tetiana-KET) [reply src/pages/Login/schema.ts]: are you sure i should replace line 18 - }); with - .regex(TEST_PASSWORD_REG_EX, 'Invalid password'),

Comment (Tetiana-KET) [reply src/pages/Login/schema.ts]: replace line 17 in IDE

**Review by JsPowWow — APPROVED** [review]

**Review by JsPowWow — APPROVED** [review]

**Pull request merged**